File: week1/MINKYOUNG/10866.cpp
// boj 10866: 덱
#include <iostream>
#include <deque>

using namespace std;

int main() {

    int N, X;
    string S;
    deque<int> d;

    cin >> N;

    for (int i=0; i<N; i++) {
        cin >> S;

        // 1. push_front X
        if (S == "push_front") {
            cin >> X;
            d.push_front(X);
        }
        // 2. push_back X
        else if (S == "push_back") {
            cin >> X;
            d.push_back(X);
        }
        // 3. pop_front
        else if (S == "pop_front")
            // 덱이 비어있지 않는 경우
            if (!d.empty()) {
                cout << d.front() << "\n";
                d.pop_front();
            // 덱이 비어 있는 경우 -1 출력
            } 
            else {
                cout << -1 << "\n";
            }
        // 4. pop_back
        else if (S == "pop_back")
            // 덱이 비어있지 않는 경ㅇ 
            if (!d.empty()) {
                cout << d.back() << "\n";
                d.pop_back();
            } 
            // 덱이 비어있는 경우 -1 출력
            else {
                cout << -1 << "\n";
            }
        // 5. size
        else if (S == "size") {
            cout << d.size() << "\n";
        }
        // 6. empty
        else if (S == "empty") {
            // cout << q.empty() << "\n";
            cout << (d.empty() ? 1 : 0) << "\n"; // 출력 명확히
        }
        // 7. front
        else if (S == "front") {
            if (!d.empty()) {
                cout << d.front() << "\n";
            } else {
                cout << -1 << "\n";
            }
        }
        // 8. back
        else if (S == "back") {
            if (!d.empty()) {
                cout << d.back() << "\n";
            } else {
                cout << -1 << "\n";
            }
        }
    }
    
    return 0;
}

File: week1/JUNGAN/10866.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        IntDeque deq = new IntDeque(10000);
        StringBuilder builder = new StringBuilder();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        for (int i = 0; i < N; i++) {
            String s = br.readLine();
            if(s.startsWith("push_front")) {
                deq.pushFront(Integer.parseInt(s.split(" ")[1]));
            }
            else if(s.startsWith("push_back")) {
                deq.pushBack(Integer.parseInt(s.split(" ")[1]));
            }
            else if(s.equals("pop_front")) {
                builder.append(deq.popFront()).append("\n");
            }
            else if(s.equals("pop_back")) {
                builder.append(deq.popBack()).append("\n");
            }
            else if(s.equals("size")) {
                builder.append(deq.size()).append("\n");
            }
            else if(s.equals("empty")) {
                builder.append(deq.isEmpty()?1:0).append("\n");
            }
            else if(s.equals("front")) {
                builder.append(deq.front()).append("\n");
            }
            else if(s.equals("back")) {
                builder.append(deq.back()).append("\n");
            }
        }
        System.out.println(builder);
    }


}

class IntDeque{
    int[] deque;
    int size, front, back, capacity;

    IntDeque(int capacity) {
        this.capacity = capacity;
        deque = new int[capacity];
        size = front = back = 0;
    }

    void pushFront(int x) {
        size++;
        front = (front - 1 + capacity) % capacity;
        deque[front] = x;
    }

    void pushBack(int x) {
        size++;
        deque[back] = x;
        back = (back + 1) % capacity;
    }

    int popFront() {
        if(size == 0) return -1;
        size--;
        int x = deque[front];
        front = (front + 1) % capacity;
        return x;
    }

    int popBack() {
        if(size == 0) return -1;
        size--;
        back = (back - 1 + capacity) % capacity;
        return deque[back];
    }

    int size(){
        return size;
    }

    boolean isEmpty(){
        return size == 0;
    }

    int front(){
        if(size == 0) return -1;
        return deque[front];
    }

    int back(){
        if(size == 0) return -1;
        return deque[(back - 1 +capacity) % capacity];
    }
}